### ISSUE TITLE
Add testing new frontend with assertions enabled

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#31739]
We won't have checked-mode anymore with the new frontend but we will need to support a modes where assertions are enabled/disabled.

Currently e.g. `pkg/vm/bin/gen_kernel.dart` - which is what the buildbot uses - does not have such a flag.

/cc @mraleph @kmillikin

**Comment by eernstg** [comment]
Here's a possible name for that flag: `--enable-asserts`. That's what both `dart` and `dart2js` have been using to enable assertions at least since 2017-07-07 (3b05eb95530b1684a3fbcc69bda1784db393b966). As an alternative, I recently head somebody suggest `--debug`, which would allow us to put more things under the same hat.

**Comment by mkustermann** [comment]
@kmillikin @mraleph @whesse 

We need proper testing of Dart 2.0 with asserts enabled (but not checked mode). This requires:
   * having `tools/test.py` have a flag, similar to the old `--checked`
   * have our individual tools accept such a flag
   * have buildbot support for those

The testing issue just came up again due to the `assert(await foo())` isssue which @kmillikin is implementing and which we need tests for.

Should we bump the priority of this?

**Comment by whesse** [comment]
Yes.  The --vm-flags option should work, if you need an immediate fix, but it would'nt be visible in the status files.  What would you like the variable to look like in the status files?
$asserts
$asserts_enabled
Would it just be for the the VM?
$vm_asserts
?

**Comment by kmillikin** [comment]
It's not just for the VM.  dart2js has --enable-asserts as a compile-time option.  As far as I know, DDC always compiles them but there are TODOs in the code that indicate that they might support an option to disable them in the future.

Ideally we would standardize the name of the flag for all our tools and then the status file variable would be the name of the flag so you never have to guess what it is.  I suggest $enable_asserts unless we can support a hyphen in the name in which case it should be $enable-asserts.

**Comment by whesse** [comment]
OK, I propose adding --enable-asserts to test.py and the status environment (as $enable_asserts), and passing --enable-asserts to the VM and dart2js, in those cases, unless I am told that a different flag should be passed.

**Comment by kmillikin** [comment]
SGTM.  We should also have the buildbots run tests with that flag for at least one architecture (VM) or runtime (dart2js).

You might also consider changing --checked so it doesn't imply --enable-asserts on the VM since it's now a separate option.  Or not, because the status file updates needed to make that work for all combinations might outweigh the benefits.

**Comment by athomas** [comment]
There is already a --enable-asserts option in test.py and a corresponding $enable_asserts variable.

**Comment by mkustermann** [comment]
Then maybe it's just a matter of removing the --checked flag from VM / test.py and changing the status files to use $enable_asserts.

**Comment by johnniwinther** [comment]
@mkustermann Do you know if there is more work left here?

**Comment by johnniwinther** [comment]
@whesse Do you now the current state of this? Can we close this issue?

**Comment by athomas** [comment]
The CFE unit tests are running with asserts but the SDK tests without.

I vaguely recall that the CFE was doing some rather heavy assertions that caused this to be very slow when turned on.
